File: Sample/Source/NodeFunctions/Container.py
# -*- encoding: utf-8 -*-

from typing import Any, Dict, List


@Meta(DisplayName='LOC("DICT_GET")', DisplayDesc='LOC("DICT_GET_DESC")')
@ReturnType(value=object)
def DictGet(dict_: Dict, key: Any) -> Any:
    return dict_.get(key)


@Meta(DisplayName='LOC("DICT_ADD")', DisplayDesc='LOC("DICT_ADD_DESC")')
@ExecSplit(default=(None,))
def DictAdd(dict_: Dict, key: Any, value: Any) -> None:
    dict_[key] = value


@Meta(DisplayName='LOC("DICT_REMOVE")', DisplayDesc='LOC("DICT_REMOVE_DESC")')
@ExecSplit(default=(None,))
def DictRemove(dict_: Dict, key: Any) -> None:
    dict_.pop(key)


@Meta(DisplayName='LOC("DICT_CLEAR")', DisplayDesc='LOC("DICT_CLEAR_DESC")')
@ExecSplit(default=(None,))
def DictClear(dict_: Dict) -> None:
    dict_.clear()


@Meta(DisplayName='LOC("DICT_CONTAINS")', DisplayDesc='LOC("DICT_CONTAINS_DESC")')
@ReturnType(value=bool)
def DictContains(dict_: Dict, key: Any) -> bool:
    return key in dict_


@Meta(DisplayName='LOC("LIST_GET")', DisplayDesc='LOC("LIST_GET_DESC")')
@ReturnType(value=object)
def ListGet(list_: List[Any], index: int) -> Any:
    return list_[index]


@Meta(DisplayName='LOC("LIST_APPEND")', DisplayDesc='LOC("LIST_APPEND_DESC")')
@ExecSplit(default=(None,))
def ListAppend(list_: List[Any], value: Any) -> None:
    list_.append(value)


@Meta(DisplayName='LOC("LIST_REMOVE")', DisplayDesc='LOC("LIST_REMOVE_DESC")')
@ExecSplit(default=(None,))
def ListRemove(list_: List[Any], index: int) -> None:
    list_.pop(index)


@Meta(DisplayName='LOC("LIST_FIND")', DisplayDesc='LOC("LIST_FIND_DESC")')
@ReturnType(index=int)
def ListFind(list_: List[Any], value: Any) -> int:
    return list_.index(value)


@Meta(DisplayName='LOC("LIST_CLEAR")', DisplayDesc='LOC("LIST_CLEAR_DESC")')
@ExecSplit(default=(None,))
def ListClear(list_: List[Any]) -> None:
    list_.clear()


@Meta(DisplayName='LOC("LIST_CONTAINS")', DisplayDesc='LOC("LIST_CONTAINS_DESC")')
@ReturnType(value=bool)
def ListContains(list_: List[Any], value: Any) -> bool:
    return value in list_

File: Sample/Scripts/Source/NodeFunctions/Container.lua
local Container = {}

---@generic T
---@param list  T[]
---@param index integer
---@return integer
local function listIndex(list, index)
    index = index or 0
    if index < 0 then
        index = #list + index
    end
    if index < 0 or index >= #list then
        error("list index out of range")
    end
    return index + 1
end

function Container.ForLoop(firstIndex, lastIndex, step)
    firstIndex = firstIndex == nil and 0 or firstIndex
    lastIndex = lastIndex == nil and 0 or lastIndex
    step = step == nil and 1 or step
    return firstIndex, lastIndex, step
end

function Container.ForEach(list_)
    return list_
end

function Container.CreateDict()
    return {}
end

function Container.DictGet(dict_, key)
    if Class.isInstance(dict_, dict) then
        return dict_:get(key)
    end
    return dict_[key]
end

function Container.DictAdd(dict_, key, value)
    dict_[key] = value
end

function Container.DictRemove(dict_, key)
    if Class.isInstance(dict_, dict) then
        dict_:pop(key)
        return
    end
    if dict_[key] == nil then
        error("dictionary key not found: " .. tostring(key))
    end
    dict_[key] = nil
end

function Container.DictClear(dict_)
    if Class.isInstance(dict_, dict) then
        dict_:clear()
        return
    end
    for key in pairs(dict_) do
        dict_[key] = nil
    end
end

function Container.DictContains(dict_, key)
    if Class.isInstance(dict_, dict) then
        return dict_:contains(key)
    end
    return dict_[key] ~= nil
end

function Container.TableToDict(table_)
    return dict(table_)
end

function Container.DictToTable(dict_)
    return dict_:toTable()
end

function Container.CreateList()
    return {}
end

function Container.ListGet(list_, index)
    return list_[listIndex(list_, index)]
end

function Container.ListAppend(list_, value)
    if Class.isInstance(list_, list) then
        list_:append(value)
        return
    end
    list_[#list_ + 1] = value
end

function Container.ListExtend(list_, values)
    if Class.isInstance(list_, list) then
        list_:extend(values)
        return
    end
    local length = #values
    for index = 1, length do
        list_[#list_ + 1] = values[index]
    end
end

function Container.ListRemove(list_, index)
    local resolvedIndex = listIndex(list_, index)
    if Class.isInstance(list_, list) then
        list_:pop(resolvedIndex)
        return
    end
    table.remove(list_, resolvedIndex)
end

function Container.ListFind(list_, value)
    local index
    if Class.isInstance(list_, list) then
        if list_:contains(value) then
            index = list_:index(value)
        end
    else
        index = table.index(list_, value)
    end
    if index ~= nil then
        return index - 1
    end
    error(tostring(value) .. " is not in list")
end

function Container.ListClear(list_)
    if Class.isInstance(list_, list) then
        list_:clear()
        return
    end
    for index = #list_, 1, -1 do
        list_[index] = nil
    end
end

function Container.ListContains(list_, value)
    if Class.isInstance(list_, list) then
        return list_:contains(value)
    end
    return table.contains(list_, value)
end

function Container.TableToList(table_)
    return list(table_)
end

function Container.ListToTable(list_)
    return list_:toTable()
end

return Container
